app/upgrades/v4/upgrades.go

```diff
@@ -1,27 +1,210 @@
+	}
+	_ = Accounts
+	_ = TotalUbedrock
+	_ = UbedrockDistribute
 )
 
 func CreateUpgradeHandler(
 	mm *module.Manager,
 	configurator module.Configurator,
-	bankKeeper bankKeeper.Keeper,
+	bankKeeper bankkeeper.Keeper,
+	accKeeper *authkeeper.AccountKeeper,
+	staking *stakingkeeper.Keeper,
 ) upgradetypes.UpgradeHandler {
 	return func(ctx sdk.Context, plan upgradetypes.Plan, fromVM module.VersionMap) (module.VersionMap, error) {
 		// logger := ctx.Logger()
 
-		// if types.IsMainnet(ctx.ChainID()) {
-		// TODO: Logic upgadeHandler
-		// }
+		if types.IsMainnet(ctx.ChainID()) {
+			// TODO: Logic upgadeHandler
+			bankBaseKeeper, _ := bankKeeper.(bankkeeper.BaseKeeper)
+			BurnToken(ctx, types.StakingCoinDenom, accKeeper, &bankBaseKeeper, staking)
+			BurnToken(ctx, types.StripeCoinDenom, accKeeper, &bankBaseKeeper, staking)
+			MintUbedrockForInitialAccount(ctx, &bankBaseKeeper, staking)
+		}
 		return mm.RunMigrations(ctx, configurator, fromVM)
 	}
 }
 
 // TODO: Function helper for upgradeHandler
+// Burn denom token
+func BurnToken(ctx sdk.Context, denom string, accKeeper *authkeeper.AccountKeeper, bank *bankkeeper.BaseKeeper, staking *stakingkeeper.Keeper) {
+	if denom == types.StakingCoinDenom {
+		// Get coin back from validator
+		accounts := accKeeper.GetAllAccounts(ctx)
+		for _, acc := range accounts {
+			_, ok := acc.(*authtypes.ModuleAccount)
+			if !ok {
+				GetbackCoinFromVal(ctx, acc.GetAddress(), staking)
+			}
+		}
+	}
+
+	// Get all account balances
+	accs := bank.GetAccountsBalances(ctx)
+	for _, acc := range accs {
+		balance := acc.Coins.AmountOf(denom)
+		// Check if ubedrock amount GT 0
+		if balance.GT(math.ZeroInt()) {
+			amount := sdk.NewCoin(denom, balance)
+			// Send ubedrock to module
+			err := bank.SendCoinsFromAccountToModule(ctx, sdk.MustAccAddressFromBech32(acc.Address), types.PaymentsProcessorName, sdk.NewCoins(amount))
+			if err != nil {
+				panic(err)
+			}
+			// Burn ubedrock in module
+			err = bank.BurnCoins(ctx, types.PaymentsProcessorName, sdk.NewCoins(amount))
+			if err != nil {
+				panic(err)
+			}
+		}
+	}
+}
+
+// Mint ubedrock for 8 security account
+func MintUbedrockForInitialAccount(ctx sdk.Context, bank *bankkeeper.BaseKeeper, staking *stakingkeeper.Keeper) {
+	// Mint coin for module
+	err := bank.MintCoins(ctx, types.PaymentsProcessorName, sdk.NewCoins(sdk.NewCoin(types.StakingCoinDenom, TotalUbedrock)))
+	if err != nil {
+		panic(err)
+	}
+
+	// Send coin from module to account
+	for _, acc := range Accounts {
+		err = bank.SendCoinsFromModuleToAccount(
+			ctx,
+			types.PaymentsProcessorName,
+			sdk.MustAccAddressFromBech32(acc),
+			sdk.NewCoins(sdk.NewCoin(types.StakingCoinDenom, UbedrockDistribute[acc])),
+		)
+		if err != nil {
+			panic(err)
+		}
+	}
+
+	// Send 1 ubedrock to each validator
+	vals := staking.GetAllValidators(ctx)
+	for _, val := range vals {
+		_, err = staking.Delegate(
+			ctx,
+			sdk.MustAccAddressFromBech32(EngineHotWal),
+			math.OneInt(),
```

Comment:
it has to be 1000000 ubedrock = 1 bedrock

Reply:
fixed